crates/matrix-sdk-base/src/event_cache/store/memory_store.rs

```diff
@@ -148,6 +148,35 @@ impl EventCacheStore for MemoryStore {
         Ok(())
     }
 
+    async fn filter_duplicated_events(
+        &self,
+        room_id: &RoomId,
+        mut events: Vec<OwnedEventId>,
+    ) -> Result<Vec<OwnedEventId>, Self::Error> {
+        // Collect all duplicated events.
+        let inner = self.inner.read().unwrap();
+
+        let mut duplicated_events = Vec::new();
+
+        for event in inner.events.unordered_events(room_id) {
+            // If `events` is empty, we can short-circuit.
+            if events.is_empty() {
+                break;
+            }
+
+            if let Some(event_id_a) = event.event_id() {
+                // This event exists in the store event!
+                if let Some(position) =
+                    events.iter().position(|event_id_b| &event_id_a == event_id_b)
+                {
+                    duplicated_events.push(events.remove(position));
+                }
```

Comment:
Maybe we could name `event_id_a` a `known_event_id` and `event_id_b` a `new_event_id`?

Reply:
Done!